File: backend/src/services/chatService.ts
import { getMongoDb } from '../config/database';
import { ChatMessage } from '../types';
import { logger } from '../utils/logger';

export class ChatService {
  async saveMessage(message: ChatMessage): Promise<void> {
    try {
      const db = getMongoDb();
      await db.collection('messages').insertOne({
        ...message,
        timestamp: new Date(),
        is_deleted: false,
      });
    } catch (error) {
      logger.error('Error saving chat message:', error);
    }
  }

  async getRecentMessages(streamId: string, limit: number = 50): Promise<ChatMessage[]> {
    try {
      const db = getMongoDb();
      const messages = await db
        .collection('messages')
        .find({ stream_id: streamId, is_deleted: false })
        .sort({ timestamp: -1 })
        .limit(limit)
        .toArray();

      return messages.reverse();
    } catch (error) {
      logger.error('Error fetching chat messages:', error);
      return [];
    }
  }

  async deleteMessage(messageId: string): Promise<void> {
    try {
      const db = getMongoDb();
      await db.collection('messages').updateOne(
        { _id: messageId },
        { $set: { is_deleted: true } }
      );
    } catch (error) {
      logger.error('Error deleting chat message:', error);
    }
  }

  async getMessageCount(streamId: string): Promise<number> {
    try {
      const db = getMongoDb();
      const count = await db
        .collection('messages')
        .countDocuments({ stream_id: streamId, is_deleted: false });
      return count;
    } catch (error) {
      logger.error('Error getting message count:', error);
      return 0;
    }
  }
}

export const chatService = new ChatService();

File: backend-go/internal/service/chat_service.go
package service

import (
	"context"
	"time"

	"github.com/phitonias/streak/internal/database"
	"github.com/phitonias/streak/internal/models"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo/options"
)

type ChatService struct{}

func NewChatService() *ChatService {
	return &ChatService{}
}

func (s *ChatService) SaveMessage(message *models.ChatMessage) error {
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	collection := database.MongoDB.Collection("messages")
	_, err := collection.InsertOne(ctx, message)
	return err
}

func (s *ChatService) GetRecentMessages(streamID string, limit int) ([]models.ChatMessage, error) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	collection := database.MongoDB.Collection("messages")

	opts := options.Find().
		SetSort(bson.D{{Key: "timestamp", Value: -1}}).
		SetLimit(int64(limit))

	cursor, err := collection.Find(ctx, bson.M{
		"stream_id":  streamID,
		"is_deleted": false,
	}, opts)

	if err != nil {
		return nil, err
	}
	defer cursor.Close(ctx)

	var messages []models.ChatMessage
	if err := cursor.All(ctx, &messages); err != nil {
		return nil, err
	}

	// Reverse to get chronological order
	for i, j := 0, len(messages)-1; i < j; i, j = i+1, j-1 {
		messages[i], messages[j] = messages[j], messages[i]
	}

	return messages, nil
}

func (s *ChatService) DeleteMessage(messageID string) error {
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	collection := database.MongoDB.Collection("messages")
	_, err := collection.UpdateOne(
		ctx,
		bson.M{"_id": messageID},
		bson.M{"$set": bson.M{"is_deleted": true}},
	)

	return err
}

func (s *ChatService) GetMessageCount(streamID string) (int64, error) {
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	collection := database.MongoDB.Collection("messages")
	count, err := collection.CountDocuments(ctx, bson.M{
		"stream_id":  streamID,
		"is_deleted": false,
	})

	return count, err
}
